### PR TITLE
Migrate PointCloud display

### PR DESCRIPTION
**Pull request by anhosi**
This is the migration of the PointCloud display without any refactorings (these will be submitted with a second PR).
The pluginlib workaround is still in place and needs to be removed once the pluginlib is available.

For testing purposes a trivial publisher node can be found in the `pointcloud_publisher` folder.

### CODE REVIEW

File: pointcloud_publisher/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,26 @@
+cmake_minimum_required(VERSION 3.5)
+project(pointcloud_publisher)
```
Comment (wjwwood) [inline pointcloud_publisher/CMakeLists.txt]: I'd rather not include this new package into the repository. Can it just be integrated into the `rviz_common` package temporarily or something?

You could also accomplish this with a python script or with the `ros2 topic echo` command line tool.

File: rviz_common/include/rviz_common/selection/selection_handler.hpp [hunk 2]
```diff
@@ -41,9 +41,9 @@
 
 #include <OgreMovableObject.h>
 
-#include "../interactive_object.hpp"
+#include "src/rviz_common/interactive_object.hpp"
 #include "rviz_common/properties/property.hpp"
-#include "../viewport_mouse_event.hpp"
+#include "src/rviz_common/viewport_mouse_event.hpp"
```
Comment (wjwwood) [inline rviz_common/include/rviz_common/selection/selection_handler.hpp]: This doesn't look right. A public header cannot reference a file in the src folder.

File: rviz_common/package.xml [hunk 3]
```diff
@@ -20,11 +20,13 @@
   <depend>rviz_rendering</depend>
 
   <build_depend>qtbase5-dev</build_depend>
+  <build_depend>rcl</build_depend>
 
   <exec_depend>libqt5-core</exec_depend>
   <exec_depend>libqt5-gui</exec_depend>
   <exec_depend>libqt5-opengl</exec_depend>
   <exec_depend>libqt5-widgets</exec_depend>
+  <exec_depend>rcl</exec_depend>
```
Comment (wjwwood) [inline rviz_common/package.xml]: Why is this needed?

File: rviz_common/src/rviz_common/selection/selection_handler.cpp [hunk 4]
```diff
@@ -54,7 +54,7 @@
 #include "rviz_common/logging.hpp"
 
 #include "rviz_common/properties/property.hpp"
-#include "../selection/selection_manager.hpp"
+#include "include/rviz_common/selection/selection_manager.hpp"
```
Comment (wjwwood) [inline rviz_common/src/rviz_common/selection/selection_handler.cpp]: This looks wrong, should not include the `include`.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp [hunk 5]
```diff
@@ -0,0 +1,1038 @@
...
+    scene_node_ = 0;
+  }
+}
+
+PointCloudCommon::PointCloudCommon(rviz_common::Display * display)
+//: spinner_(1, &cbqueue_)
+: auto_size_(false),
+  new_xyz_transformer_(false),
+  new_color_transformer_(false)
+//, transformer_class_loader_(NULL)
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: Please remove or add a TODO which describes what needs to be done.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp [hunk 6]
```diff
@@ -0,0 +1,1038 @@
...
+    SIGNAL(requestOptions(rviz_common::properties::EnumProperty *)),
+    this, SLOT(setColorTransformerOptions(rviz_common::properties::EnumProperty *)));
+}
+
+void PointCloudCommon::initialize(
+  rviz_common::DisplayContext * context,
+  Ogre::SceneNode * scene_node)
+{
+//  transformer_class_loader_ = new pluginlib::ClassLoader<PointCloudTransformer>( "rviz",
+// "rviz::PointCloudTransformer" );
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: I cannot commit to this pr, so could you add:

```
// TODO(wjwwood): reenable this when pluginlib is available
```

So I have an easier time finding it.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp [hunk 7]
```diff
@@ -0,0 +1,1038 @@
...
+
+  context_ = context;
+  scene_node_ = scene_node;
+
+  updateStyle();
+  updateBillboardSize();
+  updateAlpha();
+  updateSelectable();
+
+//  spinner_.start();
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: Please add:

```
// TODO(wjwwood): replace with executor when callback groups can be added to executors
```

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp [hunk 8]
```diff
@@ -0,0 +1,1038 @@
...
+}
+
+PointCloudCommon::~PointCloudCommon()
+{
+//  spinner_.stop();
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: Please add:

```
// TODO(wjwwood): replace with executor when callback groups can be added to executors
```

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp [hunk 9]
```diff
@@ -0,0 +1,1038 @@
...
+//    info.transformer->createProperties( display_, PointCloudTransformer::Support_XYZ,
+// info.xyz_props );
+//    setPropertiesHidden( info.xyz_props, true );
+//
+//    info.transformer->createProperties( display_, PointCloudTransformer::Support_Color,
+// info.color_props );
+//    setPropertiesHidden( info.color_props, true );
+//
+//    transformers_[ name ] = info;
+//  }
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.cpp:477]: Please add a TODO for this block of commented code.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp [hunk 10]
```diff
@@ -0,0 +1,1038 @@
...
+        color_transformer_property_->setStringStd(valid_color.rbegin()->second);
+      }
+    }
+  }
+}
+
+void PointCloudCommon::updateStatus()
+{
+  std::stringstream ss;
+  // ss << "Showing [" << total_point_count_ << "] points from [" << clouds_.size() << "] messages";
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: This seems wrong, since it will update the state to empty?

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp [hunk 11]
```diff
@@ -0,0 +1,1038 @@
...
+  if (!cloud_info->scene_node_) {
+//    if (!context_->getFrameManager()->getTransform(cloud_info->message_->header,
+// cloud_info->position_, cloud_info->orientation_))
+//    {
+//      std::stringstream ss;
+//      ss << "Failed to transform from frame [" << cloud_info->message_->header.frame_id << "]
+// to frame [" << context_->getFrameManager()->getFixedFrame() << "]";
+//      display_->setStatusStd(rviz_common::properties::StatusProperty::Error, "Message", ss.str());
+//      return false;
+//    }
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: Please add a TODO.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.hpp [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+# include <queue>
+# include <vector>
+# include <string>
+
+# include <QObject>  // NOLINT
+# include <QList>  // NOLINT
+
+//# include <message_filters/time_sequencer.h>
+
+//# include <pluginlib/class_loader.h>
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.hpp]: Please add:

```
// TODO(wjwwood): revist file when pluginlib is available
```

File: rviz_common/src/rviz_default_plugins/point_cloud_common.hpp [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+# include <map>
+# include <memory>
+# include <queue>
+# include <vector>
+# include <string>
+
+# include <QObject>  // NOLINT
+# include <QList>  // NOLINT
+
+//# include <message_filters/time_sequencer.h>
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.hpp:45]: Can you add a TODO here which describes when this can be used again or how it should change.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.hpp [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+  void initialize(rviz_common::DisplayContext * context, Ogre::SceneNode * scene_node);
+
+  void fixedFrameChanged();
+  void reset();
+  void update(float wall_dt, float ros_dt);
+
+  void addMessage(const sensor_msgs::msg::PointCloud::ConstSharedPtr & cloud);
+  void addMessage(const sensor_msgs::msg::PointCloud2::ConstSharedPtr & cloud);
+
+//  ros::CallbackQueueInterface* getCallbackQueue() { return &cbqueue_; }
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.hpp]: TODO or remove please.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.hpp [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+    PointCloudTransformerPtr trans,
+    std::string name,
+    const std::string & lookup_name);
+
+  float getSelectionBoxSize();
+  void setPropertiesHidden(const QList<rviz_common::properties::Property *> & props, bool hide);
+  void fillTransformerOptions(rviz_common::properties::EnumProperty * prop, uint32_t mask);
+
+//  ros::AsyncSpinner spinner_;
+//  ros::CallbackQueue cbqueue_;
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_common.hpp]: TODO or remove, you can add me as the TODO person and mention something about the executors and callback groups so I can find it again.

File: rviz_common/src/rviz_default_plugins/point_cloud_display.cpp [hunk 16]
```diff
@@ -27,36 +27,38 @@
...
+      "Advanced: set the size of the incoming PointCloud message queue. "
+      " Increasing this is useful if your incoming TF data is delayed significantly "
+      "from your PointCloud data, but it can greatly increase memory usage if the "
+      "messages are big.",
+      this, SLOT(updateQueueSize()));
 
   // PointCloudCommon sets up a callback queue with a thread for each
   // instance.  Use that for processing incoming messages.
-  update_nh_.setCallbackQueue( point_cloud_common_->getCallbackQueue() );
+//  update_nh_.setCallbackQueue( point_cloud_common_->getCallbackQueue() );
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_display.cpp]: Please add a TODO here for me:

```
// TODO(wjwwood): replace with executor when callback groups can be added to executors
```

File: rviz_common/src/rviz_default_plugins/point_cloud_transformer.hpp [hunk 17]
```diff
@@ -27,39 +27,48 @@
...
+#ifndef RVIZ_DEFAULT_PLUGINS__POINT_CLOUD_TRANSFORMER_HPP_
+#define RVIZ_DEFAULT_PLUGINS__POINT_CLOUD_TRANSFORMER_HPP_
 
-#include <QObject>
+#include <vector>
 
-#include <ros/message_forward.h>
+#include <QObject>  // NOLINT
+
+// #include <ros/message_forward.h>
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_transformer.hpp]: remove or TODO

File: rviz_common/src/rviz_default_plugins/point_cloud_transformers.cpp [hunk 18]
```diff
@@ -0,0 +1,632 @@
...
+
+  int32_t xi = findChannelIndex(cloud, "x");
+  int32_t yi = findChannelIndex(cloud, "y");
+  int32_t zi = findChannelIndex(cloud, "z");
+
+  const uint32_t xoff = cloud->fields[xi].offset;
+  const uint32_t yoff = cloud->fields[yi].offset;
+  const uint32_t zoff = cloud->fields[zi].offset;
+  const uint32_t point_step = cloud->point_step;
+//  const uint32_t num_points = cloud->width * cloud->height; // TODO(greimela): Unused?
```
Comment (wjwwood) [inline rviz_common/src/rviz_default_plugins/point_cloud_transformers.cpp]: If the variable is unused, feel free to remove it.

File: rviz_rendering/include/rviz_rendering/point_cloud.hpp [hunk 19]
```diff
@@ -27,8 +27,14 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
-#ifndef OGRE_TOOLS_OGRE_POINT_CLOUD_H
-#define OGRE_TOOLS_OGRE_POINT_CLOUD_H
+#ifndef RVIZ_RENDERING__POINT_CLOUD_HPP_
+#define RVIZ_RENDERING__POINT_CLOUD_HPP_
+
+#include <stdint.h>
```
Comment (wjwwood) [inline rviz_rendering/include/rviz_rendering/point_cloud.hpp:33]: Use `cstdint` instead.

File: rviz_rendering/include/rviz_rendering/point_cloud.hpp [hunk 20]
```diff
@@ -58,31 +58,39 @@ class RenderSystem;
...
 
-  Ogre::RenderOperation* getRenderOperation() { return &mRenderOp; }
+#ifdef __clang__
+# pragma clang diagnostic push
+# pragma clang diagnostic ignored "-Woverloaded-virtual"
+#endif
+  Ogre::RenderOperation * getRenderOperation() {return &mRenderOp;}
+#ifdef __clang__
+# pragma clang diagnostic pop
+#endif
```
Comment (wjwwood) [inline rviz_rendering/include/rviz_rendering/point_cloud.hpp:78]: This looks like a warning you should not suppress, it looks like the signature has changed. Might be something new with 1.10.

File: rviz_rendering/include/rviz_rendering/point_cloud.hpp [hunk 21]
```diff
@@ -164,47 +172,47 @@ class PointCloud : public Ogre::MovableObject
...
   void regenerateAll();
   void shrinkRenderables();
 
   Ogre::AxisAlignedBox bounding_box_;       ///< The bounding box of this point cloud
   float bounding_radius_;                   ///< The bounding radius of this point cloud
 
   typedef std::vector<Point> V_Point;
-  V_Point points_;                          ///< The list of points we're displaying.  Allocates to a high-water-mark.
+  V_Point points_;                          ///< The list of points we're displaying.
+  ///< Allocates to a high-water-mark.
```
Comment (wjwwood) [inline rviz_rendering/include/rviz_rendering/point_cloud.hpp]: I don't this comment should go after the variable it is describing, instead move the whole trailing comment to the line before the member variable.

File: rviz_rendering/src/rviz_rendering/point_cloud.cpp [hunk 22]
```diff
@@ -551,17 +500,16 @@ void PointCloud::addPoints(Point* points, uint32_t num_points)
       current_vertex_count = 0;
 
       vertex_size = op->vertexData->vertexDeclaration->getVertexSize(0);
-      fptr = (float*)((uint8_t*)vdata);
+      fptr = reinterpret_cast<float *>(reinterpret_cast<uint8_t *>(vdata));
```
Comment (wjwwood) [inline rviz_rendering/src/rviz_rendering/point_cloud.cpp:503]: Not 100% sure about this one, I think either the interior reinterpret cast is not needed or it should be a static cast?

**Review by wjwwood — CHANGES_REQUESTED** [review]
I made a lot of requests for you to add TODO's for me (I would make them but I cannot commit here).

Also, in general, if a piece of code was not already commented out, but is after the diff then it needs to be removed or it needs a comment/TODO describing when it can be removed, readded, or changed.

File: rviz_rendering/src/rviz_rendering/point_cloud.cpp (hunk 22 above)
Comment (Martin-Idel-SI) [reply rviz_rendering/src/rviz_rendering/point_cloud.cpp:503]: The interior cast is actually completely unnecessary. This will be addressed in the follow-up PR we are preparing where we do the refactoring. For now, this was just meant to shut up the linter. But if you like, we can of course edit it here already.

File: rviz_rendering/include/rviz_rendering/point_cloud.hpp (hunk 20 above)
Comment (Martin-Idel-SI) [reply rviz_rendering/include/rviz_rendering/point_cloud.hpp:78]: This warning is about name-hiding. There is (and always has been) a method `void getRenderOperation(Ogre::RenderOperation &)` in `Ogre::SimpleRenderable`. Since the name of the method is the same, this overloading here hides the original function. It seems that this has always been the case, but neither gcc nor msvc complain about it.

File: rviz_rendering/include/rviz_rendering/point_cloud.hpp (hunk 19 above)
Comment (Martin-Idel-SI) [reply rviz_rendering/include/rviz_rendering/point_cloud.hpp:33]: Will be addressed in the refactoring PR.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp (hunk 10 above)
Comment (Martin-Idel-SI) [reply rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: This was already commented out in the original rviz code. It seems odd, but I don't understand what the correct behaviour should be - I'll add a TODO and we revisit during refactoring.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp (hunk 7 above)
Comment (Martin-Idel-SI) [reply rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: Would this be necessary at all? I would actually remove it for now, since everything works fine with the executors we added in the visualization manager.

File: pointcloud_publisher/CMakeLists.txt (hunk 1 above)
Comment (Martin-Idel-SI) [reply pointcloud_publisher/CMakeLists.txt]: We agree. We'll delete the commit introducing this for now. Later, when we want to add visual testing, maybe we'll need it - but we can discuss this at that point.

**Comment by Martin-Idel-SI** [comment]
We're currently revisiting all new files, so a lot will be improved in the near future. But for now, I hope my changes suffice.

File: rviz_rendering/include/rviz_rendering/point_cloud.hpp (hunk 20 above)
Comment (wjwwood) [reply rviz_rendering/include/rviz_rendering/point_cloud.hpp:78]: Ok, sounds like a false positive then 👍.

At best we could let the Ogre guys know that they should use a different name for virtual functions.

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp (hunk 10 above)
Comment (wjwwood) [reply rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: Sounds good, I couldn't tell the difference in the diff (GitHub didn't see them as rename and change).

File: rviz_common/src/rviz_default_plugins/point_cloud_common.cpp (hunk 7 above)
Comment (wjwwood) [reply rviz_common/src/rviz_default_plugins/point_cloud_common.cpp]: Well, this code was originally expecting to have its own thread, so to emulate that we would need the changes to the executor I was planning. Maybe we don't end up needing it, but I think it might be important.

**Comment by wjwwood** [comment]
I made a few follow up commits, but otherwise I think it lgtm. We can merge so I can rebase my stuff and you can make your follow up pr.

Thanks for splitting this work up!

**Review by wjwwood — APPROVED** [review]
